data/tutorials/tt_02_file_manipulation.md

```diff
@@ -9,123 +9,618 @@ date: 2021-05-27T21:07:30-00:00
+[dedicated tutorial](/docs/error-handling)
+
+### Catching Exceptions
+
+In OCaml, many of the file and directory operations can raise an exception when
+they encounter an error. For example, trying to open a non-existent file with
+`In_channel.open_text` will raise a `Sys_error` exception. Therefore, it's a
+good practice to catch these exceptions and handle them accordingly.
+
+Let's see how we can handle exceptions when reading from a file:
+
+```ocaml
+let read_from_file filename =
+  try
+    let ic = In_channel.open_text filename in
+    let content = In_channel.input_all ic in
+    In_channel.close ic;
+    content
+  with
+  | Sys_error msg -> print_endline ("Could not read file: " ^ msg); ""
+```
+
+In this function, we use a `try ... with` block to catch any `Sys_error` that
+might be raised when opening the file, reading its content, or closing it. If
+such an exception is raised, we print an error message and return an empty
+string.
+
+### Checking Beforehand
+
+Another way to avoid exceptions is to check beforehand whether an operation is
+likely to succeed. For example, before opening a file, you can check whether it
+exists and whether you have the necessary permissions to open it:
+
+```ocaml
+let read_from_file filename =
+  if Sys.file_exists filename then
+    let ic = In_channel.open_text filename in
+    let content = In_channel.input_all ic in
+    In_channel.close ic;
+    content
+  else
+    print_endline ("File does not exist: " ^ filename); ""
+```
+
+This function checks whether the file exists before trying to open it. If the
+file does not exist, it prints an error message and returns an empty string.
+
+## Common Pitfalls
+
+This section covers some frequent pitfalls when working with files in OCaml and
+provides remedies to prevent them.
+
+### Ensuring Immediate Writes: Flushing `out_channels`
+
+OCaml buffers `out_channel` data and only writes it to the file when the buffer
+is full or when the channel is closed. If immediate write to a file is needed,
+you must explicitly flush the `out_channel` using the flush function.
+
+Here's an example:
+
+```ocaml
+let oc = open_out "myfile.txt"
+output_string oc "Hello, world!"
+flush oc
+```
+
+### Remembering to Close Channels
+
+When you open a file, the operating system allocates a file descriptor for it.
+File descriptors are a limited resource, so it's important to release them when
+you're done using them by closing the file. Furthermore, not closing an
+`out_channel` might result in data loss, because some data might still be in the
+buffer and not yet written to the file.
+
+However, simply closing the file at the end of the function is not sufficient,
+because if an exception is raised before the function reaches the point where it
+closes the file, the file will remain open. You should therefore ensure that
+files are always closed, even if an exception is raised. One way to do this is
+to use a `try ... with` block:
```

Comment:
`try .. with` blocks are not good enough in this case and should be replaced by `Fun.protect`.
Also, the good practice is to use `In_channel.with_open_text`, which remove this concern.